pkg/txn/storage/memorystorage/memtable/tree_diff_iter.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2022 Matrix Origin
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memtable
+
+import (
+	"database/sql"
+
+	"github.com/tidwall/btree"
+)
+
+var _ DiffIterPiece[dumbRow, dumbRow] = new(Tree[dumbRow, dumbRow, dumbRow])
+
+func (t *Tree[K, V, R]) NewDiffIter(fromTime, toTime *Time) DiffIter[K, V] {
+	ret := &TreeDiffIter[K, V]{
+		fromTime: fromTime,
+		toTime:   toTime,
+		iter:     t.rows.Copy().Iter(),
+	}
+	return ret
+}
+
+type TreeDiffIter[
+	K Ordered[K],
+	V any,
+] struct {
+	fromTime *Time
+	toTime   *Time
+	iter     btree.GenericIter[*PhysicalRow[K, V]]
+
+	currentRow      *PhysicalRow[K, V]
+	currentVersion  Version[V]
+	currentIsDelete bool
+}
+
+func (t *TreeDiffIter[K, V]) Close() error {
+	t.iter.Release()
+	return nil
+}
+
+func (t *TreeDiffIter[K, V]) Seek(pivot K) bool {
+	if !t.iter.Seek(&PhysicalRow[K, V]{
+		Key: pivot,
+	}) {
+		return false
+	}
+	return t.read()
+}
+
+func (t *TreeDiffIter[K, V]) First() bool {
+	if !t.iter.First() {
+		return false
+	}
+check:
```

Comment:
What does the loop really do?

Reply:
The first entry in the tree may not contains a valid diff item in the range, the loop skips these entries before a valid one.
I'll send changes in this PR in separate PRs.